src/cares_wrap.cc

```diff
@@ -525,6 +544,573 @@ class QueryWrap : public AsyncWrap {
+  hostent* host;
+
+  int status;
+  switch (*type) {
+  case ns_t_a:
+  case ns_t_cname:
+  case ns_t_cname_or_a:
+    status = ares_parse_a_reply(buf,
+                                len,
+                                &host,
+                                reinterpret_cast<ares_addrttl*>(addrttls),
+                                naddrttls);
+    break;
+  case ns_t_aaaa:
+    status = ares_parse_aaaa_reply(buf,
+                                   len,
+                                   &host,
+                                   reinterpret_cast<ares_addr6ttl*>(addrttls),
+                                   naddrttls);
+    break;
+  case ns_t_ns:
+    status = ares_parse_ns_reply(buf, len, &host);
+    break;
+  case ns_t_ptr:
+    status = ares_parse_ptr_reply(buf, len, NULL, 0, AF_INET, &host);
+    break;
+  default:
+    CHECK(0 && "Bad NS type");
+  }
+
+  if (status != ARES_SUCCESS)
+    return status;
+
+  /* If it's `CNAME`, return the CNAME value */
+  /* And if it's `CNAME_OR_A` and it has value in `h_name` and `h_aliases[0]` */
+  /* We consider it's a CNAME record, otherwise we consider it's an A record */
+  if ((*type == ns_t_cname_or_a && host->h_name && host->h_aliases[0]) ||
+      *type == ns_t_cname) {
+    // A cname lookup always returns a single record but we follow the
+    // common API here.
+    *type = ns_t_cname;
+    ret->Set(ret->Length(), OneByteString(env->isolate(), host->h_name));
+    ares_free_hostent(host);
+    return ARES_SUCCESS;
+  }
+
+  if (*type == ns_t_cname_or_a)
+    *type = ns_t_a;
+
+  if (*type == ns_t_ns) {
+    HostentToNames(env, host, ret);
+  } else if (*type == ns_t_ptr) {
+    uint32_t offset = ret->Length();
+    for (uint32_t i = 0; host->h_aliases[i] != NULL; i++) {
+      ret->Set(i + offset, OneByteString(env->isolate(), host->h_aliases[i]));
+    }
+  } else {
+    HostentToAddresses(env, host, ret);
+  }
+
+  ares_free_hostent(host);
+
+  return ARES_SUCCESS;
+}
+
+
+int ParseMxReply(Environment* env,
+                 unsigned char* buf,
+                 int len,
+                 Local<Array> ret,
+                 bool need_type = false) {
+  HandleScope handle_scope(env->isolate());
+
+  struct ares_mx_reply* mx_start;
+  int status = ares_parse_mx_reply(buf, len, &mx_start);
+  if (status != ARES_SUCCESS) {
+    return status;
+  }
+
+  Local<String> exchange_symbol = env->exchange_string();
+  Local<String> priority_symbol = env->priority_string();
+  Local<String> type_symbol = env->type_string();
+  Local<String> mx_symbol = env->dns_mx_string();
+
+  uint32_t offset = ret->Length();
+  ares_mx_reply* current = mx_start;
+  for (uint32_t i = 0; current != nullptr; ++i, current = current->next) {
+    Local<Object> mx_record = Object::New(env->isolate());
+    mx_record->Set(exchange_symbol,
+                   OneByteString(env->isolate(), current->host));
+    mx_record->Set(priority_symbol,
+                   Integer::New(env->isolate(), current->priority));
```

Comment:
Regarding all uses of `Set()`: Can you use the overload that returns a `Maybe<bool>` and check the result?